### ISSUE TITLE
Please document how to generate fitz.py and don't include it in the release tarball and/or git

### ISSUE THREAD

**Issue by josch** [pymupdf/PyMuPDF#312]
Hi,

I'm currently [packaging PyMuPDF for Debian](https://salsa.debian.org/python-team/modules/pymupdf) but there are a few remaining problems before we can upload it into the archive.

The biggest problem is `fitz/fitz.py` which is autogenerated from `*.i` files using SWIG but SWIG is not run as part of running `setup.py build` and the resulting generated file is even part of the release tarball. This creates multiple problems for downstreams like us.

Firstly, `fitz/fitz.py` is not source in the sense that it is not a file that you will end up editing manually to improve PyMuPDF. Thus, every contributor has to somehow figure out how to rebuild `fitz/fitz.py` from source using SWIG and manually launch the respective tools. By integrating SWIG into the build process, you automate this step and make it easier for users of your software to contribute.

Secondly, since `fitz/fitz.py` is autogenerated, it should not be part of the git repository. The revision control system should only include the sources themselves and not any autogenerated files. The autogenerated files will just clutter your commit diffs and take up unnecessary space. If you think that it's important to ship autogenerated files to your users, then you can add `fitz/fitz.py` to your release tarballs.

Thirdly, suppose there is a security update to SWIG. Are you keeping track of those, update your SWIG install and immediately make a new PyMuPDF build with the newly generated `fitz/fitz.py`? By relying on your downstreams to generate `fitz.py` using SWIG you help making sure that even older releases of PyMuPDF do get built with an updated SWIG version. This argument does not only hold for security updates but also any other relevant SWIG updates.

In summary, having `fitz/fitz.py` tracked by git is similar to keeping files generated by autotools in git or keeping a pdf generated from LaTeX in git. By keeping it, you make it harder for downstreams like Debian who do care about the security and freedom related implications because to make sure that we don't ship any autogenerated files, we will have to manually remove any autogenerated files from your releases. It would be more convenient for us and probably also for others if autogenerated files were not part of what you ship in the first place.

Thanks!

cheers, josch

**Comment by JorjMcKie** [comment]
Hi,
If I am following your arguments correctly, then **neither** ``fitz.py`` **nor** ``fitz_wrap.c`` should be part of the installation material.
Shouldn't be a problem to get rid of them with the next version.

It is more problematic to overhaul the installation procedure such that SWIG generation becomes a part of it. It would imply creating installation alternatives for (1) people having (resp. willing to install) SWIG and (2) the others.
On the other hand it's my impression that the vast majority of users nowadays cares for the wheels only anyway. So an installation using sources could well be **forced** to using SWIG. However, I don't like the idea to fight multi-front battles for consequential issues on using SWIG on each of the zillion different platforms ... Maybe I just include a How-To chapter and of course provide the necessary material.
Let me think about it.

As per the SWIG release:
Yes, I am regularly checking for new versions - without necessarily triggering a new PyMuPDF version just because of SWIG. Version 3.0.12 (which I am using since its publication in January 2017) has also been stable for more than two years. So this hasn't ever been a practical issue.
The next PyMuPDF version will however be based on SWIG version 4.0.
Following your request, the SWIG version will in future only be relevant for binary (wheels) installations. Source installers (like you) will have to decide themselves, which SWIG version to use. PyMuPDF itself currently does not impose a specific one.

Cheers,
Jorj

**Comment by JorjMcKie** [comment]
Hi,
If I am following your arguments correctly, then **neither** ``fitz.py`` **nor** ``fitz_wrap.c`` should be part of the installation material.
Shouldn't be a problem to get rid of them with the next version.

It is more problematic to overhaul the installation procedure such that SWIG generation becomes a part of it. It would imply creating installation alternatives for (1) people having (resp. willing to install) SWIG and (2) the others.
On the other hand it's my impression that the vast majority of users nowadays cares for the wheels only anyway. So an installation using sources could well be **forced** to using SWIG. However, I don't like the idea to fight multi-front battles for consequential issues on using SWIG on each of the zillion different platforms ... Maybe I just include a How-To chapter and of course provide the necessary material.
Let me think about it.

As per the SWIG release:
Yes, I am regularly checking for new versions - without necessarily triggering a new PyMuPDF version just because of SWIG. Version 3.0.12 (which I am using since its publication in January 2017) has also been stable for more than two years. So this hasn't ever been a practical issue.
The next PyMuPDF version will however be based on SWIG version 4.0.
Following your request, the SWIG version will in future only be relevant for binary (wheels) installations. Source installers (like you) will have to decide themselves, which SWIG version to use. PyMuPDF itself currently does not impose a specific one.

Cheers,
Jorj

**Comment by JorjMcKie** [comment]
Hi,
If I am following your arguments correctly, then **neither** ``fitz.py`` **nor** ``fitz_wrap.c`` should be part of the installation material.
Shouldn't be a problem to get rid of them with the next version.

It is more problematic to overhaul the installation procedure such that SWIG generation becomes a part of it. It would imply creating installation alternatives for (1) people having (resp. willing to install) SWIG and (2) the others.
On the other hand it's my impression that the vast majority of users nowadays cares for the wheels only anyway. So an installation using sources could well be **forced** to using SWIG. However, I don't like the idea to fight multi-front battles for consequential issues on using SWIG on each of the zillion different platforms ... Maybe I just include a How-To chapter and of course provide the necessary material.
Let me think about it.

As per the SWIG release:
Yes, I am regularly checking for new versions - without necessarily triggering a new PyMuPDF version just because of SWIG. Version 3.0.12 (which I am using since its publication in January 2017) has also been stable for more than two years. So this hasn't ever been a practical issue.
The next PyMuPDF version will however be based on SWIG version 4.0.
Following your request, the SWIG version will in future only be relevant for binary (wheels) installations. Source installers (like you) will have to decide themselves, which SWIG version to use. PyMuPDF itself currently does not impose a specific one.

Cheers,
Jorj

**Comment by JorjMcKie** [comment]
Hi,
If I am following your arguments correctly, then **neither** ``fitz.py`` **nor** ``fitz_wrap.c`` should be part of the installation material.
Shouldn't be a problem to get rid of them with the next version.

It is more problematic to overhaul the installation procedure such that SWIG generation becomes a part of it. It would imply creating installation alternatives for (1) people having (resp. willing to install) SWIG and (2) the others.
On the other hand it's my impression that the vast majority of users nowadays cares for the wheels only anyway. So an installation using sources could well be **forced** to using SWIG. However, I don't like the idea to fight multi-front battles for consequential issues on using SWIG on each of the zillion different platforms ... Maybe I just include a How-To chapter and of course provide the necessary material.
Let me think about it.

As per the SWIG release:
Yes, I am regularly checking for new versions - without necessarily triggering a new PyMuPDF version just because of SWIG. Version 3.0.12 (which I am using since its publication in January 2017) has also been stable for more than two years. So this hasn't ever been a practical issue.
The next PyMuPDF version will however be based on SWIG version 4.0.
Following your request, the SWIG version will in future only be relevant for binary (wheels) installations. Source installers (like you) will have to decide themselves, which SWIG version to use. PyMuPDF itself currently does not impose a specific one.

Cheers,
Jorj

**Comment by JorjMcKie** [comment]
Hi,
If I am following your arguments correctly, then **neither** ``fitz.py`` **nor** ``fitz_wrap.c`` should be part of the installation material.
Shouldn't be a problem to get rid of them with the next version.

It is more problematic to overhaul the installation procedure such that SWIG generation becomes a part of it. It would imply creating installation alternatives for (1) people having (resp. willing to install) SWIG and (2) the others.
On the other hand it's my impression that the vast majority of users nowadays cares for the wheels only anyway. So an installation using sources could well be **forced** to using SWIG. However, I don't like the idea to fight multi-front battles for consequential issues on using SWIG on each of the zillion different platforms ... Maybe I just include a How-To chapter and of course provide the necessary material.
Let me think about it.

As per the SWIG release:
Yes, I am regularly checking for new versions - without necessarily triggering a new PyMuPDF version just because of SWIG. Version 3.0.12 (which I am using since its publication in January 2017) has also been stable for more than two years. So this hasn't ever been a practical issue.
The next PyMuPDF version will however be based on SWIG version 4.0.
Following your request, the SWIG version will in future only be relevant for binary (wheels) installations. Source installers (like you) will have to decide themselves, which SWIG version to use. PyMuPDF itself currently does not impose a specific one.

Cheers,
Jorj

**Comment by josch** [comment]
Yes, you are correct. Neither `fitz.py` nor `fitz_wrap.c` should be part of the sources because they are not the "source" but something that gets generated from a source.

I have never used it but it seems that Python distutils understands swig. So if you pass the `*.i` files as sources into the distutils `Extension`, then it will automatically call `swig -python ...`.

Which command do you usually use to generate `fitz.py` and `fitz_wrap.c`? Maybe I can create a pull request for you.

As for the release, my suggestion would be to either make a source release which only includes the sources as well as a binary release (wheels). Users of wheels don't have to worry about SWIG anyways. Users of the source distribution will need SWIG but that is already the case right now. Even today, somebody who wants to build PyMuPDF from source so that they can edit and improve it, will need SWIG. By incorporating SWIG into `setup.py` I argue that you would actually make it easier for contributors because the `setup.py` would contain the full recipe of how to get from the original source to a final binary.

**Comment by JorjMcKie** [comment]
I would definitely welcome any such PR :-) ... while I do agree with your argument:

> I argue that you would actually make it easier for contributors because the setup.py would contain the full recipe of how to get from the original source to a final binary.

But as you can see from the bulk of comments in that ``setup.py`` file, there are  lots of details changing by platform. And a source installation of PyMuPDF is admittedly not the easiest one - because of the external MuPDF dependency (1) as such, and (2) the requirement to adjust ``config.h`` in there, and (3) copy over some corrective sources into it before MuPDF can even be generated.

All this generates **many, MANY** "opportunities" for bugs. So I think I will ever only **describe** what someone must do for a source generation and not take over responsibility for code actually **doing** it. But having a collection of proven source installation procedures for more and more platforms would definitely be great!

As a (mainly) Windows guy, for generating PyMuPDF I currently use this procedure:

1. execute SWIG simply like this: ``swig -python fitz.i``
2. compile the ``fitz_wrap.c`` output using all my 10 Python installations (versions 3.7, 3.4, 3.5, 3.6, 3.7, each for 32bit and 64bit) using Visual Studio in batch.
3. Generate the Windows wheels and ZIP installations on my machine
4. Generate Linux and Mac OSX wheels using Travis (64bit only)
5. Install and test the Linux and Windows wheels on my machine. OSX is not tested, because I have no Mac.
6. Publish Git release and upload to PIP.

As can be seen, I am not using ``setup.py`` for the source installation myself (any longer).

**Comment by josch** [comment]
A related question to packaging PyMuPDF for Debian: are your three patches to mupdf upstreamable? In other words: have they been forwarded to the mupdf developers and if not, why?

**Comment by JorjMcKie** [comment]
They are not of the same categories:
* ``config.h`` is **intended** to be modified by source installers (ie. you and me). This version contains my recommended values. Using the original will give the generated extension module a size of beyond 32 MB b/o included support of myriads of ``TOFU`` fonts.
* ``error.c`` is my modification of a MuPDF module to meet some feature request. The original in MuPDF has no errors and will ever stay as it is. Not including my version will cause unpleasant messages appear on stderr. So a cosmetical feature if you will.
* ``pdf_device.c`` contains a bug which has been fixed for MuPDF v1.15.0, so will go away from PyMuPDF distribution.

So the only indispensable is the last one in v1.14.xx.

**Comment by josch** [comment]
Thanks for your rationale! Did you ever consider submitting your changes to `error.c` to the mupdf developers or do you have reason to believe that they would not accept that change? It looks harmless to me.

**Comment by JorjMcKie** [comment]
Hm ... looking for some tactful wording. Maybe this is appropriate:

Artifex (MuPDF) is a profit oriented company. They entertain MuPDF per its permissive license obviously in the hope to sell commercial licenses (expensive!). So everything not being a bug but just a sidestep, is treated with great reluctance.

My change also has a somewhat worse performance - not really relevant because we are not talking about even kilobytes of data, but ...

You should also know that they have a homegrown javascript interpreter (freeware & open source) with builtin MuPDF support - in some perspective quite similar to PyMuPDF ... so who knows which other motives come into play.

**Comment by josch** [comment]
Just as a datapoint: As I was working on integrating swig with setup.py, I noticed that my swig version (3.0.12) replaced `strncpy` with `memcpy` in `fitz/fitz_wrap.c`. Due to the subtle difference between these two functions, this shows that changes in swig *do* have security implications.

**Comment by josch** [comment]
For Debian, the patch to build with swig is simple:

```patch
--- a/setup.py
+++ b/setup.py
@@ -5,7 +5,7 @@ import sys, os
 if sys.platform.startswith("linux"):
     module = Extension(
         "fitz._fitz",  # name of the module
-        ["fitz/fitz_wrap.c"],  # C source file
+        ["fitz/fitz.i"],
         include_dirs=[  # we need the path of the MuPDF and zlib headers
             "/usr/include/mupdf",
             "/usr/local/include/mupdf",
```

Python setuptools will do the rest. I didn't test any other platforms besides Debian GNU/Linux.

**Comment by josch** [comment]
Another question concerning packaging: Which license is your code released under? `setup.py` says 'GPL3+' but `nano_setup.py` and `PKG-INFO` say `GPL3+ and AGPL3+`. I know that the resulting binary is licensed under both licenses due to the mupdf license. But which license is the *source* licensed as?

**Comment by josch** [comment]
Pull request building `fitz.py` and `fitz_wrap.c` via `setup.py`: https://github.com/pymupdf/PyMuPDF/pull/314

**Comment by JorjMcKie** [comment]
> 
> 
> Just as a datapoint: As I was working on integrating swig with setup.py, I noticed that my swig version (3.0.12) replaced `strncpy` with `memcpy` in `fitz/fitz_wrap.c`. Due to the subtle difference between these two functions, this shows that changes in swig _do_ have security implications.

Sure there are differences.

When migrating to SWIG 4,0 I also noticed that ``fitz.i`` must receive minor adjustments to still generate a valid ``fitz.py`` due to SWIG code changes w/r to object creation.

Still testing here.

**Comment by JorjMcKie** [comment]
> 
> 
> Another question concerning packaging: Which license is your code released under? `setup.py` says 'GPL3+' but `nano_setup.py` and `PKG-INFO` say `GPL3+ and AGPL3+`. I know that the resulting binary is licensed under both licenses due to the mupdf license. But which license is the _source_ licensed as?

Can't recall where this ``nano_setup.py`` came from. Was contributed by a user and I forgot who and why. I guess we can get rid of it.

To clarify:

* Every component coming from Artifex / MuPDF is AGPL3+. So this extends to their sources and the resulting binaries. It also applies to my mods of ``config.h``, ``error.c`` and ``pdf_device.c``.
* **Everything else** is GPL3+. ``setup.py`` reads ``PGO-INFO`` and makes the long description from it, so it shouldn't introduce any differences here.
* The generated ``fitz.py`` / ``fitz_wrap.c`` also are GPL3+ ... if SWIG doesn't introduce another complication, which I am not 100% certain about. The resulting binary ``_fitz.so`` should be AGPL3+ (the more restrictive compared to GPL3+) because it contains code from MuPDF.

**Comment by josch** [comment]
Thanks for your clarifications! Then maybe you should adjust the `PKG-INFO` file because that one says:

```
Classifier: License :: OSI Approved :: GNU Affero General Public License v3 or later (AGPLv3+)
Classifier: License :: OSI Approved :: GNU General Public License v3 or later (GPLv3+)
```

Maybe it also lists AGPL3+ because there are AGPL3+ bits in there, namely the `.c` and `.h` files from mupdf?

But okay, I'll then list all files as being `GPL3+`. I only found few files with copyright headers saying otherwise, namely:

```
Files: demo/PDFdisplay.py
Copyright: 2015 Jorj X. McKie <jorj.x.mckie@outlook.de>
License: GPL-3 and AGPL-3

Files: demo/curly-polygon.py
       demo/draw-sines.py
       demo/fitz-logo.py
       demo/svg-logo.py
       examples/4-up.py
       examples/PDFLinkMaint.py
       examples/ParseTab.py
       examples/colordbHSV.py
       examples/colordbRGB.py
       examples/csv2meta.py
       examples/csv2toc.py
       examples/embedded-copy.py
       examples/embedded-export.py
       examples/embedded-import.py
       examples/embedded-list.py
       examples/meta2csv.py
       examples/posterize.py
       examples/wxTableExtract.py
       examples/wxTableExtract.py
Copyright: 2015 Jorj X. McKie <jorj.x.mckie@outlook.de>
License: GPL-3
```

Note the difference between `GPL-3` in above files and `GPL-3+` the remaining codebase is licensed under. Maybe this was just an oversight and the above files should actually all be `GPL-3+` as well?

Also, please tell me if I should open separate issues for my queries instead of hijacking this one or whether you'd be more comfortable continuing our conversation by email.

Thanks!

cheers, josch

**Comment by JorjMcKie** [comment]
Thank you for your interest and contributions!
Rest assured that I do appreciate these open ends are getting straightened out.

Maybe e-mail is more practicable, so please feel free to use mine.

The above examples should all be of the same GPL3+ - I haven't been careful enough pulling this through.

I have changed ``setup.py`` meanwhile and removed the separate ``license=`` argument in the hope that then the PKG-INFO specification will exclusively prevail.

**Comment by JorjMcKie** [comment]
> Then maybe you should adjust the PKG-INFO file because that one says ...

I am afraid I cannot do that, because it is supplied for PyMuPDF as a whole - including the binary versions (wheels and Windows ZIP files).

**Comment by josch** [comment]
Ah okay, yes then you are of course correct and then the `PKG-INFO` must continue to list GPL3+ as well as AGPL3+. I now recorded that all *source* files from the PyMuPDF git repository except for `fitz/_error.c`, `fitz/_mupdf_config.h`, `fitz/_pdf-device.c`, `fitz/fitz.py` and `fitz/fitz_wrap.c` (the latter two recently being deleted) are licensed under the terms of GPL-3 or later -- even if the header in some of the files says otherwise (see above for the list of discrepancies I found). Thanks!

**Comment by JorjMcKie** [comment]
> Ah okay, yes then you are of course correct and then the `PKG-INFO` must continue to list GPL3+ as well as AGPL3+. I now recorded that all _source_ files from the PyMuPDF git repository except for `fitz/_error.c`, ``

Agreed!
I looked up the SWIG license document (again) in the meantime and found confirmed, that SWIG **_output_** stays under the license of the **_input_** - ie. GPL3+ in our case.

**Comment by josch** [comment]
PyMuPDF is now waiting to be reviewed by Debian FTP Masters: https://ftp-master.debian.org/new/pymupdf_1.14.16-1.html

Thus, closing this issue. Thanks for your help! :)
